File: lilac/src/main/java/com/hartwig/hmftools/lilac/variant/SomaticCodingCount.java
package com.hartwig.hmftools.lilac.variant;

import com.hartwig.hmftools.common.variant.CodingEffect;
import com.hartwig.hmftools.lilac.hla.HlaAllele;

import java.util.List;
import java.util.stream.Collectors;

public class SomaticCodingCount
{
    public final HlaAllele Allele;

    private double mInframeIndel;
    private double mMissense;
    private double mNonsense;
    private double mSplice;
    private double mSynonymous;

    public SomaticCodingCount(final HlaAllele allele, double inframeIndel, double missense, double nonsense, double splice, double synonymous)
    {
        Allele = allele;
        mInframeIndel = inframeIndel;
        mMissense = missense;
        mNonsense = nonsense;
        mSplice = splice;
        mSynonymous = synonymous;
    }

    public double inframeIndel() { return mInframeIndel; }
    public double missense() { return mMissense; }
    public double nonsense() { return mNonsense; }
    public double splice() { return mSplice; }
    public double synonymous() { return mSynonymous; }

    public String toString()
    {
        return "SomaticCodingCount(allele=" + Allele + ", inframeIndel=" + mInframeIndel + ", missense=" + mMissense
                + ", nonsense=" + mNonsense + ", splice=" + mSplice + ", synonymous=" + mSynonymous + ")";
    }

    public double total() { return mInframeIndel + mMissense + mNonsense + mSplice + mSynonymous; }

    public static List<SomaticCodingCount> create(final List<HlaAllele> winners)
    {
        return winners.stream()
                .map(x -> new SomaticCodingCount(x, 0, 0, 0, 0, 0)).collect(Collectors.toList());
    }

    public static void addVariant(
            final List<SomaticCodingCount> codingCounts, final SomaticVariant variant, final List<HlaAllele> variantAlleles)
    {
        boolean isIndel = variant.Alt.length() != variant.Ref.length();
        CodingEffect codingEffect = variant.CanonicalCodingEffect;
        addVariant(codingCounts, isIndel, codingEffect, variantAlleles);
    }

    public static void addVariant(
            final List<SomaticCodingCount> codingCounts, boolean isIndel, final CodingEffect effect, final List<HlaAllele> variantAlleles)
    {
        double contribution = 1.0 / variantAlleles.size();

        for(HlaAllele allele : variantAlleles)
        {
            // give only to the first matching allele
            SomaticCodingCount matchingCount = codingCounts.stream().filter(x -> x.Allele.equals(allele)).findFirst().orElse(null);

            if(matchingCount != null)
                matchingCount.addVariant(isIndel, effect, contribution);
        }
    }

    private void addVariant(boolean indel, CodingEffect effect, double contribution)
    {
        if(indel && effect == CodingEffect.MISSENSE)
        {
            mInframeIndel += contribution;
            return;
        }

        switch(effect)
        {
            case MISSENSE:
                mMissense += contribution;
                break;

            case NONSENSE_OR_FRAMESHIFT:
                mNonsense += contribution;
                break;

            case SPLICE:
                mSplice += contribution;
                break;

            case SYNONYMOUS:
                mSynonymous += contribution;
                break;

            default:
                break;
        }
    }

}

File: lilac/src/main/kotlin/com/hartwig/hmftools/lilac/variant/SomaticCodingCount.kt
package com.hartwig.hmftools.lilac.variant

import com.hartwig.hmftools.common.variant.CodingEffect
import com.hartwig.hmftools.common.variant.VariantContextDecorator
import com.hartwig.hmftools.lilac.hla.HlaAllele

data class SomaticCodingCount(val allele: HlaAllele, val inframeIndel: Double, val missense: Double, val nonsense: Double, val splice: Double, val synonymous: Double) {
    val total = inframeIndel + missense + nonsense + splice + synonymous

    companion object {
        fun create(winners: List<HlaAllele>): List<SomaticCodingCount> {
            return winners.sorted().map { SomaticCodingCount(it, 0.0, 0.0, 0.0, 0.0, 0.0) }
        }

        fun List<SomaticCodingCount>.addVariant(variant: VariantContextDecorator, variantAlleles: Set<HlaAllele>): List<SomaticCodingCount> {
            return this.addVariant(variant.alt().length != variant.ref().length, variant.canonicalCodingEffect(), variantAlleles);
        }

        fun List<SomaticCodingCount>.addVariant(indel: Boolean, effect: CodingEffect, variantAlleles: Set<HlaAllele>): List<SomaticCodingCount> {
            val contribution = 1.0 / variantAlleles.size
            val result = mutableListOf<SomaticCodingCount>()
            result.addAll(this.filter { it.allele !in variantAlleles })

            for (variantAllele in variantAlleles) {
                val counts = this.filter { it.allele == variantAllele }.sortedBy { it.total }

                // Split between them
                // result.addAll(counts.map { it.addVariant(variantEffect, contribution / counts.size) })

                // Alternative, give only to first to first
                if (counts.isNotEmpty()) {
                    result.add(counts[0].addVariant(indel, effect, contribution))
                    result.addAll(counts.takeLast(counts.size - 1))
                }
            }

            return result.sortedBy { it.allele }
        }
    }

    private fun addVariant(indel: Boolean, effect: CodingEffect, contribution: Double): SomaticCodingCount {
        if (indel && effect == CodingEffect.MISSENSE) {
            return copy(inframeIndel = inframeIndel + contribution)
        }

        return when (effect) {
            CodingEffect.MISSENSE -> copy(missense = missense + contribution)
            CodingEffect.NONSENSE_OR_FRAMESHIFT -> copy(nonsense = nonsense + contribution)
            CodingEffect.SPLICE -> copy(splice = splice + contribution)
            CodingEffect.SYNONYMOUS -> copy(synonymous = synonymous + contribution)
            else -> this
        }
    }

}